crates/hir-ty/src/lib.rs

```diff
@@ -15,6 +15,9 @@ extern crate rustc_abi;
 #[cfg(not(feature = "in-rust-tree"))]
 extern crate ra_ap_rustc_abi as rustc_abi;
 
+// Don't use the in-tree one even when we could.
+extern crate ra_ap_rustc_pattern_analysis as rustc_pattern_analysis;
+
```

Comment:
What is the problem with the in tree one?

Reply:
I'd like to be able to update them separately. E.g. if I make a breaking change in the in-tree one, I don't need to fix rust-analyzer instantly. It just adds a lot of complexity compared to treating it like a normal crate.

Reply:
Said differently, the reason we use the in-tree rustc-lexer etc is because it's important to follow closely the rustc version. There's no such need for exhaustiveness checking

Reply:
I'll rephrase the comment actually

Reply:
It's fine for now that things are moving fast, but when the dust settles I think it would be better to use the in-tree version here as well. Using the in-tree version ensures that r-a can keep its rustc dependencies update without worrying about api breakages in the rustc side.

Reply:
You mean it's a way for the rustc side to be responsible for not breaking r-a? Ok that makes sense. For now I'm taking that responsibility by hand, and yeah when the dust settles we can automate it

Reply:
Yes, that's one of the proposes of this in-tree thing.

Using the in-tree version also has the benefit that if the pull request is completely unrelated to the r-a and just fixes the breakages, you can do r-a changes in the same PR in the rust repository, which may make things simpler (1 PR instead of 2 PRs).

Reply:
Hm, that's a strong point. Right now I'd have to wait for the rustc one to be merged, wait for republish, and make a PR to r-a.

What happens when a PR that changes r-a is merged into rustc? Does it get merged into r-a master automatically?

Reply:
Subtree syncs happen every week, but they often fail because of bootstrap shenanigans.

Reply:
Hm, slow either way. Does rustc CI run rust-analyzer tests?

Reply:
Ok, then for now I'd rather not use the in-tree crate, that will make it easier to iterate